src/plugins/fs/udf.c

```diff
@@ -0,0 +1,494 @@
+    g_autofree gchar *vid = NULL;
+
+    if (extra) {
+        for (extra_p = extra; *extra_p; extra_p++)
+            g_ptr_array_add (options_array, bd_extra_arg_copy ((BDExtraArg *) *extra_p));
+    }
+
+    if (options->label) {
+        /* there are different limits for lvid and vid and we need to set both so lets just
+           truncate vid (GParted does the same thing) */
+        if (g_str_is_ascii (options->label) && strlen (options->label) > 30)
+            vid = g_strndup (options->label, 30);
+        else if (g_utf8_validate (options->label, -1, NULL) && g_utf8_strlen (options->label, -1) > 15)
+            vid = g_utf8_substring (options->label, 0, 15);
+        else
+            vid = g_strdup (options->label);
+
+        g_ptr_array_add (options_array, bd_extra_arg_new ("--lvid", options->label));
+        g_ptr_array_add (options_array, bd_extra_arg_new ("--vid", vid));
+    }
+
+    if (options->uuid)
+        g_ptr_array_add (options_array, bd_extra_arg_new ("-u", options->uuid));
+
+    g_ptr_array_add (options_array, NULL);
+
+    return (BDExtraArg **) g_ptr_array_free (options_array, FALSE);
+}
+
+static gint get_blocksize (const gchar *device, GError **error) {
+    gint fd = -1;
+    gint blksize = 0;
+
+    fd = open (device, O_RDONLY);
+    if (fd < 0) {
+        g_set_error (error, BD_FS_ERROR, BD_FS_ERROR_FAIL,
+                    "Failed to open the device '%s' to get its block size", device);
+        return -1;
+    }
+
+    if (ioctl (fd, BLKSSZGET, &blksize) < 0) {
+        g_set_error (error, BD_FS_ERROR, BD_FS_ERROR_FAIL,
+                     "Failed to get block size of the device '%s'", device);
+        close (fd);
+        return -1;
+    }
+
+    close (fd);
+
+    return blksize;
+}
+
+/**
+ * bd_fs_udf_mkfs:
+ * @device: the device to create a new UDF fs on
+ * @extra: (allow-none) (array zero-terminated=1): extra options for the creation (right now
+ *                                                 passed to the 'mkudffs' utility)
+ * @error: (out): place to store error (if any)
+ *
+ * Note: This functions by default calls 'mkudffs' with '--media-type=hd --udfrev=0x201
+ *       --blocksize=<detect>' to make sure older versions of mkudffs work correctly.
+ *       If you specify additional functions additional options using the @extra
+ *       argument, these options are NOT added to avoid conflicts so make sure to add
+ *       these too if necessary.
+ *
+ * Returns: whether a new UDF fs was successfully created on @device or not
+ *
+ * Tech category: %BD_FS_TECH_UDF-%BD_FS_TECH_MODE_MKFS
+ */
+gboolean bd_fs_udf_mkfs (const gchar *device, const BDExtraArg **extra, GError **error) {
+    const gchar *args[7] = {"mkudffs", "--utf8", NULL, NULL, NULL, NULL, NULL};
+    gint blocksize = -1;
+    g_autofree gchar *bs_str = NULL;
+
+    if (!check_deps (&avail_deps, DEPS_MKUDFFS_MASK, deps, DEPS_LAST, &deps_check_lock, error))
+        return FALSE;
+
+    if (!extra) {
```

Comment:
UDF is first filesystem where we need to specify some extra arguments to make it work. These "defaults" are from UDisks and I'm not sure if we want to allow users to specify their own values and how. Options:

- Add extra arguments to `bd_fs_udf_mkfs` -- I don't like this, I would like to keep all `bd_fs_xxx_mkfs` function the same.
- Add second mkfs function, someting like `bd_fs_udf_mkfs_full`
- Do what I did here -- we have some defaults and user can override them with the `extra` argument, but needs to specify everything manually.

Reply:
For the first point is there a strong reason to keep argument consistency with other filesystems now that we have the generic `bd_fs_mkfs()`? Adding an argument at least for the `--media-type` might still benefit from the rest of the _"automatic"_ options. I.e. to change just the media type the caller wouldn't need to have knowledge of the argument structure.

The reason why I personally consider media type (and anonymous vs. full ownerships) important is a wide usage of UDF for DVD and BD media where ISO9660 + RockRidge don't support files over 2GB. However I have no idea how exactly the burning softwarez do create the UDF filesystem and what options do they provide. I could still imagine this use case for both `libblockdev` and `udisks`.

However, I'm also fine with the suggestion \#2.

At the end combination of \#2 and \#3 seems like a good way forward, if you insist on \#1.

Reply:
I changed `bd_fs_udf_mkfs`, it now takes `media_type`, `revision` and `block_size` as arguments. `bd_fs_mkfs` made it more complicated because it adds everything as extra arguments so it overwrites our defaults and changing `bd_fs_udf_mkfs` now seems to be lesser evil than changing `bd_fs_mkfs` behaviour or adding some more notes to the documentation. I still don't like it, but it's probably better this way.